sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala

```diff
@@ -50,17 +51,21 @@ case class Mode(
   override def inputTypes: Seq[AbstractDataType] = Seq(AnyDataType)
 
   override def checkInputDataTypes(): TypeCheckResult = {
-    if (UnsafeRowUtils.isBinaryStable(child.dataType) || child.dataType.isInstanceOf[StringType]) {
+    // TODO: SPARK-49358: Mode expression for map type with collated fields
+    if (UnsafeRowUtils.isBinaryStable(child.dataType) ||
+      !child.dataType.existsRecursively(f => f.isInstanceOf[MapType] &&
+        !UnsafeRowUtils.isBinaryStable(f))) {
       /*
         * The Mode class uses collation awareness logic to handle string data.
-        * Complex types with collated fields are not yet supported.
+        * All complex types except MapType with collated fields are supported.
        */
-      // TODO: SPARK-48700: Mode expression for complex types (all collations)
       super.checkInputDataTypes()
     } else {
-      TypeCheckResult.TypeCheckFailure("The input to the function 'mode' was" +
-        " a type of binary-unstable type that is " +
-        s"not currently supported by ${prettyName}.")
+      TypeCheckResult.DataTypeMismatch("UNSUPPORTED_MODE_DATA_TYPE",
+        messageParameters =
+          Map("child" -> child.dataType.simpleString,
+            "mode" -> prettyName,
+            "reason" -> "MapType with collated fields"))
```

Comment:
Is there any other `reason`s? Please, embed it to the error message tamplate.

Reply:
That is the only reason I can think of.

Reply:
I can replace it with "unsupported data type for collation-aware mode" if you want that alternate phrasing?

Reply:
I would like to avoid error texts in source code because of:
- Tech/devs editors won't be able to change only `error-conditions.json`, they have to fix the source code + tests, and rebuild Spark.
- When we will translate the errors to different languages, it would be convenient to have entire error message in one place.